numpy/lib/arraypad.py

```diff
@@ -1165,191 +836,130 @@ def pad(array, pad_width, mode, **kwargs):
+                         .format(mode, unsupported_kwargs))
+
+    stat_functions = {"maximum": np.max, "minimum": np.min,
+                      "mean": np.mean, "median": np.median}
+
+    # Create array with final shape and original values
+    # (padded area is undefined)
+    padded, original_area_slice = _pad_simple(array, pad_width)
+    # And prepare iteration over all dimensions
+    # (zipping may be more readable than using enumerate)
+    axes = range(padded.ndim)
+
+    if mode == "constant":
+        values = kwargs.get("constant_values", 0)
+        values = _as_pairs(values, padded.ndim)
+        for axis, index_pair, value_pair in zip(axes, pad_width, values):
+            roi = _view_roi(padded, original_area_slice, axis)
+            _set_pad_area(roi, axis, index_pair, value_pair)
+
+    elif mode == "empty":
+        pass  # Do nothing as padded is already prepared
+
+    elif array.size == 0:
+        # Only modes "constant" and "empty" can extend empty axes, all other
+        # modes depend on `array` not being empty
+        # -> ensure every empty axis is only "padded with 0"
+        for axis, index_pair in zip(axes, pad_width):
+            if array.shape[axis] == 0 and any(index_pair):
+                raise ValueError(
+                    "can't extend empty axis {} using modes other than "
+                    "'constant' or 'empty'".format(axis)
+                )
+        # _pad_simple already returned the correct result
+
+    elif mode == "edge":
+        for axis, index_pair in zip(axes, pad_width):
+            roi = _view_roi(padded, original_area_slice, axis)
+            edge_pair = _get_edges(roi, axis, index_pair)
+            _set_pad_area(roi, axis, index_pair, edge_pair)
+
+    elif mode == "linear_ramp":
+        end_values = kwargs.get("end_values", 0)
+        end_values = _as_pairs(end_values, padded.ndim)
+        for axis, index_pair, value_pair in zip(axes, pad_width, end_values):
+            roi = _view_roi(padded, original_area_slice, axis)
+            ramp_pair = _get_linear_ramps(roi, axis, index_pair, value_pair)
+            _set_pad_area(roi, axis, index_pair, ramp_pair)
+
+    elif mode in stat_functions:
+        func = stat_functions[mode]
+        length = kwargs.get("stat_length", None)
+        length = _as_pairs(length, padded.ndim, as_index=True)
+        for axis, index_pair, length_pair in zip(axes, pad_width, length):
+            roi = _view_roi(padded, original_area_slice, axis)
+            stat_pair = _get_stats(roi, axis, index_pair, length_pair, func)
+            _set_pad_area(roi, axis, index_pair, stat_pair)
+
+    elif mode in {"reflect", "symmetric"}:
+        method = kwargs.get("reflect_type", "even")
+        include_edge = True if mode == "symmetric" else False
+        for axis, (left_index, right_index) in zip(axes, pad_width):
+            if array.shape[axis] == 1 and (left_index > 0 or right_index > 0):
```

Comment:
Wait, this is OK if the mode is "odd" right? I am wondering if we should add deprecation warnings here. Also I am wondering if all of those repeat modes should really exist. But for now maybe lets not touch it.

Reply:
sorry, I meant "symmetric" of course.

Reply:
Yeah. This extra code path for singleton dimensions is only necessary for the mode "reflect". That mode excludes the edge from the reflection and would (for a singleton) dimension try to extend with an empty array. When I comment this block out, `np.pad([1], 2, "reflect")` will run endlessly. In the case of "symmetric" it doesn't hurt as the singleton dimension is always the same value.